prov/efa/src/rxr/rxr_cq.c

```diff
@@ -724,14 +724,34 @@ void rxr_cq_handle_shm_completion(struct rxr_ep *ep, struct fi_cq_data_entry *cq
 	}
 }
 
+static inline
+bool rxr_cq_need_tx_completion(struct rxr_ep *ep,
+			       struct rxr_tx_entry *tx_entry)
+
+{
+	if (tx_entry->fi_flags & RXR_NO_COMPLETION)
+		return false;
+
+	/*
+	 * ep->util_ep.tx_msg_flags is either 0 or FI_COMPLETION, depend on
+	 * whether app specfied FI_SELECTIVE_COMPLETION when binding CQ.
+	 * (ep->util_ep.tx_msg_flags was set in ofi_ep_bind_cq())
+	 *
+	 * If tx_msg_flags is 0, we only write completion when app specify
+	 * FI_COMPLETION in flags.
+	 */
+	return ep->util_ep.tx_msg_flags == FI_COMPLETION ||
+	       tx_entry->fi_flags & FI_COMPLETION;
+}
+
+
 void rxr_cq_write_tx_completion(struct rxr_ep *ep,
 				struct rxr_tx_entry *tx_entry)
 {
 	struct util_cq *tx_cq = ep->util_ep.tx_cq;
 	int ret;
 
-	if (!(tx_entry->fi_flags & RXR_NO_COMPLETION) &&
-	    ofi_need_completion(rxr_tx_flags(ep), tx_entry->fi_flags)) {
```

Comment:
Hmm, don't we need the ofi_need_completion checks too?

Reply:
ofi_need_completion check  ```(cq_flags & FI_SELECTIVE_COMPLETION)``` and ```(op_flags | FI_COMPLETION)```.

the information of FI_SELECTIVE_COMPLETION is contained in ```util_ep.tx_msg_attr```, the information of ```op_flags``` is contained in ```tx_entry->flags```, so ```rxr_cq_need_tx_completion``` covers all aspects of ```ofi_need_completion```.

Reply:
Talked to Wei offline. I was missing the fact that we were passing tx_op_flags as cq_flag, erroneously.